pkg/code-generator/schemagen/cue.go

```diff
@@ -0,0 +1,188 @@
+// TODO (sam-heilbron) This tool has a known flaw that makes it unusable with Gloo Edge protos
+// 	 There are open issues to track this and a GitHub discussion with more context:
+//	 https://github.com/cuelang/cue/discussions/944. We would like to move towards this
+//	 implementation and have added it now to compare schemas generated with this implementation
+//	 to schemas generated with other implementations to ensure consistency.
+type cueGenerator struct {
+	// The Collector used to extract imports for proto files
+	importsCollector collector.Collector
+	absoluteRoot     string
+	protoDir         string
+}
+
+// TODO (sam-heilbron) CUE deprecates OrderedMap
+// 	 We should start using ast.File and other APIs in the ast package to manipulate the schemas
+//	 This requires some more support from ast package thus leaving this as to-do
+
+// Mapping from protobuf message name to OpenApi schema
+type OpenApiSchemas map[string]*openapi.OrderedMap
+
+func NewCueGenerator(importsCollector collector.Collector, absoluteRoot string) *cueGenerator {
+	return &cueGenerator{
+		importsCollector: importsCollector,
+		absoluteRoot:     absoluteRoot,
+		protoDir:         anyvendor.DefaultDepDir,
+	}
+}
+
+func (c *cueGenerator) GetJsonSchemaForProject(project *model.Project) (map[schema.GroupVersionKind]*apiextv1beta1.JSONSchemaProps, error) {
+	protobufExtractor, err := c.getProtobufExtractorForProject(project)
+	if err != nil {
+		return nil, err
+	}
+
+	instances, err := protobufExtractor.Instances()
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert cuelang to openapi
+	openApiGenerator := &openapi.Generator{
+		// k8s structural schemas do not allow $refs, i.e. all references must be expanded
+		ExpandReferences: true,
+	}
+
+	openApiSchemas := OpenApiSchemas{}
+	built := cue.Build(instances)
+	for _, builtInstance := range built {
+		// Avoid generating openapi for irrelevant proto imports.
+		if !strings.HasSuffix(builtInstance.ImportPath, project.ProjectConfig.GoPackage) {
+			continue
+		}
+
+		if builtInstance.Err != nil {
+			return nil, err
+		}
+		if err = builtInstance.Value().Validate(); err != nil {
+			return nil, eris.Errorf("Cue instance validation failed for %s: %+v", project.ProtoPackage, err)
+		}
+		schemas, err := openApiGenerator.Schemas(builtInstance)
+		if err != nil {
+			return nil, eris.Errorf("Cue openapi generation failed for %s: %+v", project.ProtoPackage, err)
+		}
+
+		// Iterate openapi objects to construct mapping from proto message name to openapi schema
+		for _, kv := range schemas.Pairs() {
+			openApiSchemas[kv.Key] = kv.Value.(*openapi.OrderedMap)
+		}
+
+		return c.convertOpenApiSchemasToJsonSchemas(project, openApiSchemas)
+	}
+
+	return nil, nil
+}
+
+func (c *cueGenerator) getProtobufExtractorForProject(project *model.Project) (*protobuf.Extractor, error) {
+	// Collect all protobuf definitions including transitive dependencies.
+	var imports []string
```

Comment:
Should this function be inside the collector package instead of part of the cueGenerator? It looks like CollectImportsForFile does a similar thing for a specific file and this is just collect the protobuf definitions for a project?

Reply:
I tried to isolate our dependency on cue into a single file to keep it distinct from the rest of our codebase. It's unfortunate that it does something quiet similar to our collector package, but I think we need it because it generates the cue types that we then parse